File: scripts/validate.mjs
import { readFileSync, readdirSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const MODULES_DIR = join(__dirname, '..', 'modules');

const VALID_CATEGORIES = [
  'system', 'hardware', 'network', 'audio', 'power', 'time',
  'workspace', 'window', 'tray', 'weather', 'productivity', 'media', 'custom'
];

const UUID_PATTERN = /^[a-z0-9][a-z0-9-]*@[a-z0-9][a-z0-9-]*$/;
const VERSION_PATTERN = /^\d+\.\d+\.\d+(-[a-zA-Z0-9.]+)?$/;
const REPO_URL_PATTERN = /^https:\/\/github\.com\/[^/]+\/[^/]+$/;

export class ValidationError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ValidationError';
  }
}

export function validateUuid(uuid) {
  if (!uuid || uuid.length < 3 || uuid.length > 100) {
    throw new ValidationError(`UUID must be 3-100 characters: ${uuid}`);
  }
  if (!UUID_PATTERN.test(uuid)) {
    throw new ValidationError(`Invalid UUID format: ${uuid}`);
  }
}

export function validateVersion(version) {
  if (!VERSION_PATTERN.test(version)) {
    throw new ValidationError(`Invalid version format: ${version}`);
  }
}

export function validateRepoUrl(url) {
  if (!REPO_URL_PATTERN.test(url)) {
    throw new ValidationError(`Invalid repo_url (must be https://github.com/owner/repo): ${url}`);
  }
}

export function validateCategory(category) {
  if (!VALID_CATEGORIES.includes(category)) {
    throw new ValidationError(`Invalid category '${category}'. Valid: ${VALID_CATEGORIES.join(', ')}`);
  }
}

export function validateTags(tags) {
  if (tags === null || tags === undefined) {
    return;
  }
  if (!Array.isArray(tags)) {
    throw new ValidationError('tags must be an array');
  }
  if (tags.length > 10) {
    throw new ValidationError('Maximum 10 tags allowed');
  }
  const seen = new Set();
  for (const tag of tags) {
    if (typeof tag !== 'string' || tag.length < 1 || tag.length > 30) {
      throw new ValidationError(`Invalid tag: ${tag}`);
    }
    if (seen.has(tag)) {
      throw new ValidationError(`Duplicate tag: ${tag}`);
    }
    seen.add(tag);
  }
}

export function validateModule(data) {
  const required = ['uuid', 'name', 'description', 'author', 'category', 'version', 'repo_url'];
  for (const field of required) {
    if (!(field in data)) {
      throw new ValidationError(`Missing required field: ${field}`);
    }
  }

  validateUuid(data.uuid);
  validateVersion(data.version);
  validateRepoUrl(data.repo_url);
  validateCategory(data.category);
  validateTags(data.tags);

  const name = data.name || '';
  if (name.length < 3 || name.length > 50) {
    throw new ValidationError('name must be 3-50 characters');
  }
  const description = data.description || '';
  if (description.length < 10 || description.length > 500) {
    throw new ValidationError('description must be 10-500 characters');
  }
}

export function validateFile(filename, data) {
  validateModule(data);
  const expectedFilename = `${data.uuid}.json`;
  if (filename !== expectedFilename) {
    throw new ValidationError(`Filename mismatch: expected ${expectedFilename}, got ${filename}`);
  }
}

export function checkDuplicates(modules) {
  const seen = new Map();
  for (const m of modules) {
    const uuid = m.uuid;
    if (seen.has(uuid)) {
      throw new ValidationError(`Duplicate UUID '${uuid}'`);
    }
    seen.set(uuid, true);
  }
}

function loadModule(filepath) {
  const content = readFileSync(filepath, 'utf-8');
  return JSON.parse(content);
}

function validateDirectory(modulesDir) {
  const errors = [];
  const files = readdirSync(modulesDir).filter(f => f.endsWith('.json')).sort();

  if (files.length === 0) {
    console.log('No module files found');
    return 0;
  }

  const modules = [];
  for (const filename of files) {
    const filepath = join(modulesDir, filename);
    try {
      const data = loadModule(filepath);
      validateFile(filename, data);
      modules.push(data);
    } catch (e) {
      errors.push(`${filename}: ${e.message}`);
    }
  }

  try {
    checkDuplicates(modules);
  } catch (e) {
    errors.push(e.message);
  }

  if (errors.length > 0) {
    console.log('Validation errors:');
    for (const error of errors) {
      console.log(`  - ${error}`);
    }
    return 1;
  }

  console.log(`Validated ${files.length} modules successfully`);
  return 0;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  process.exit(validateDirectory(MODULES_DIR));
}

File: scripts/validate.py
#!/usr/bin/env python3
import json
import sys
import re
from pathlib import Path
from typing import Optional

SCHEMA_PATH = Path(__file__).parent.parent / "schema" / "module.schema.json"
MODULES_DIR = Path(__file__).parent.parent / "modules"

VALID_CATEGORIES = [
    "system", "hardware", "network", "audio", "power", "time",
    "workspace", "window", "tray", "weather", "productivity", "media", "custom"
]

UUID_PATTERN = re.compile(r"^[a-z0-9][a-z0-9-]*@[a-z0-9][a-z0-9-]*$")
VERSION_PATTERN = re.compile(r"^\d+\.\d+\.\d+(-[a-zA-Z0-9.]+)?$")
REPO_URL_PATTERN = re.compile(r"^https://github\.com/[^/]+/[^/]+$")


class ValidationError(Exception):
    pass


def validate_uuid(uuid: str) -> None:
    if not uuid or len(uuid) < 3 or len(uuid) > 100:
        raise ValidationError(f"UUID must be 3-100 characters: {uuid}")
    if not UUID_PATTERN.match(uuid):
        raise ValidationError(f"Invalid UUID format: {uuid}")


def validate_version(version: str) -> None:
    if not VERSION_PATTERN.match(version):
        raise ValidationError(f"Invalid version format: {version}")


def validate_repo_url(url: str) -> None:
    if not REPO_URL_PATTERN.match(url):
        raise ValidationError(f"Invalid repo_url (must be https://github.com/owner/repo): {url}")


def validate_category(category: str) -> None:
    if category not in VALID_CATEGORIES:
        raise ValidationError(f"Invalid category '{category}'. Valid: {VALID_CATEGORIES}")


def validate_tags(tags: Optional[list]) -> None:
    if tags is None:
        return
    if not isinstance(tags, list):
        raise ValidationError("tags must be an array")
    if len(tags) > 10:
        raise ValidationError("Maximum 10 tags allowed")
    seen = set()
    for tag in tags:
        if not isinstance(tag, str) or len(tag) < 1 or len(tag) > 30:
            raise ValidationError(f"Invalid tag: {tag}")
        if tag in seen:
            raise ValidationError(f"Duplicate tag: {tag}")
        seen.add(tag)


def validate_module(data: dict) -> None:
    required = ["uuid", "name", "description", "author", "category", "version", "repo_url"]
    for field in required:
        if field not in data:
            raise ValidationError(f"Missing required field: {field}")

    validate_uuid(data["uuid"])
    validate_version(data["version"])
    validate_repo_url(data["repo_url"])
    validate_category(data["category"])
    validate_tags(data.get("tags"))

    if len(data.get("name", "")) < 3 or len(data.get("name", "")) > 50:
        raise ValidationError("name must be 3-50 characters")
    if len(data.get("description", "")) < 10 or len(data.get("description", "")) > 500:
        raise ValidationError("description must be 10-500 characters")


def load_module(path: Path) -> dict:
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def validate_file(path: Path) -> list[str]:
    errors = []
    try:
        data = load_module(path)
        validate_module(data)
        expected_filename = f"{data['uuid']}.json"
        if path.name != expected_filename:
            errors.append(f"Filename mismatch: expected {expected_filename}, got {path.name}")
    except json.JSONDecodeError as e:
        errors.append(f"Invalid JSON: {e}")
    except ValidationError as e:
        errors.append(str(e))
    return errors


def check_duplicates(modules_dir: Path) -> list[str]:
    errors = []
    seen_uuids = {}
    for path in sorted(modules_dir.glob("*.json")):
        try:
            data = load_module(path)
            uuid = data.get("uuid")
            if uuid in seen_uuids:
                errors.append(f"Duplicate UUID '{uuid}' in {path.name} and {seen_uuids[uuid]}")
            else:
                seen_uuids[uuid] = path.name
        except (json.JSONDecodeError, KeyError):
            pass
    return errors


def main() -> int:
    all_errors = []

    if not MODULES_DIR.exists():
        print(f"ERROR: Modules directory not found: {MODULES_DIR}")
        return 1

    module_files = list(MODULES_DIR.glob("*.json"))
    if not module_files:
        print("No module files found")
        return 0

    for path in sorted(module_files):
        errors = validate_file(path)
        for error in errors:
            all_errors.append(f"{path.name}: {error}")

    dup_errors = check_duplicates(MODULES_DIR)
    all_errors.extend(dup_errors)

    if all_errors:
        print("Validation errors:")
        for error in all_errors:
            print(f"  - {error}")
        return 1

    print(f"Validated {len(module_files)} modules successfully")
    return 0


if __name__ == "__main__":
    sys.exit(main())
